WordPress/openverse-catalog
# Respository restructure

## Fixes


Fixes #231 by @AetherUnbound, fixes #262 by @AetherUnbound

## Description


There's no avoiding it, this is a **big PR**. A large number of files are modified, but most of those files have only been moved & had import paths changed. I'll enumerate all of the changes below, and try to include specific files that should be looked at that have more changes beyond moving/imports.

### Directory restructuring

I've moved files around in order to consolidate modules and make it easier to find things. The tests were also moved to mirror this new structure.

- Moved all `util/*` modules under `common/`
- Moved all `storage/*` modules under `common/`
- Moved provider API scripts and provider DAGs under `providers/`
- Moved all commoncrawl related code to `commoncrawl/`
- Moved other DAGs into subfolders related to function (`database/`, `maintenance/`, I didn't end up moving `common_api_workflows.py` because I wasn't sure where it should go!)

It was unclear to me why we had `common`, `util`, and `storage` as top level libraries, so I consolidated them all into `common`. If there was a rationale that I missed which makes separating these useful, let me know!

### Code changes

While doing this restructuring, I discovered that some tests had a hidden dependency; if I ran them in a different order, and unpredictable amount of tests would fail where they previously succeeded. After some tedious trial and error, I was able to track this down to **unexpected socket access**. The tests were set up with `pytest-socket` so that only tests which had explicitly enabled socket access could make `request` calls. Unfortunately, all tests after the `test_smithsonian_unit_codes.py` test circumvented this because of the pytest marker on the unit codes test! Adding socket exceptions for all of the failing tests seemed cumbersome and unhelpful, so I've removed that plugin entirely. If we want to assert that a test _shouldn't_ be accessing sockets, then we can add specific checks for said test. I don't think it's beneficial for us to block all requests across the board (at least at this time), seeing as how many requests are made for licenses alone. Instead, I've added the [`pytest-random-order` plugin](https://pypi.org/project/pytest-random-order/) and set this as the default run parameter for our CI. This will shuffle the tests so that they're run in a random order on GitHub. The plugin will report the seed it used, so we can use said seed locally if we ever encounter an issue with order-specific failures. I didn't set up a permanent configuration setting for this because we'd either have to 1) mount a `pytest.ini` file at the top level within the container along with the tests (it can't be in the test folder unless we want to run `pytest tests/` explicitly) or 2) add a new environment variable. The latter is plausible but this seemed to be the best balance between consistent behavior during development while still randomizing the checks during CI.

**TL;DR**: Tests had a hidden, order-specific requirement. The plugin that caused this has been removed and now tests are run in a random order on CI.

Other changes:

- Updated the `.airflowignore` file to respect the new structure
- Modified some recipes in the `justfile` to make multiple argument easier and run commands without having to start the `webserver` service
- Added directory structure documentation

---

I think that's it...thank you to whomever has the infinite patience necessary for making it this far :sweat_smile: :raised_hands:

## Testing Instructions

1. `just build`
1. `just test`
1. (Optional) `just run airflow dags list` to ensure all the relevant DAGs still get parsed

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
